sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala

```diff
@@ -162,6 +161,11 @@ private[sql] case class AddExchange(sqlContext: SQLContext) extends Rule[SparkPl
       def addExchangeIfNecessary(partitioning: Partitioning, child: SparkPlan): SparkPlan =
         if (child.outputPartitioning != partitioning) Exchange(partitioning, child) else child
 
+      // Check if the ordering we want to ensure is the same as the child's output
+      // ordering. If so, we do not need to add the Sort operator.
+      def addSortIfNecessary(ordering: Seq[SortOrder], child: SparkPlan): SparkPlan =
+        if (child.outputOrdering != ordering) Sort(ordering, global = false, child) else child
```

Comment:
The problem with doing it this way is you are no longer taking advantage of the external sort that spark can do as part of the shuffle anyway.  I think we need to holistically consider whether there needs to be a sort as part of when there needs to be a shuffle.  For now, to simplify things I think it would even be okay to always shuffle if there needs to be a sort.